Add optional report sheet showing collection coverage

himdel opened the pull request:
Issue: AAP-41887

(Started from #100, kudos to @AlexSCorey)

This adds a new optional "Data collection status" sheet to the CCSPv2 report,
enabled by `METRICS_UTILITY_OPTIONAL_CCSP_REPORT_SHEETS=data_collection_status,...`.

The data is based on `data_collection_status.csv` output by the collector.

It lists collection instances (when data was actually gathered), highlighting:
* unusual gaps between collections (based on `collection_start_timestamp`)
* gaps in collected intervals (based on `since` vs `until`)

---

Also, a bit of cleanup/refactoring in reports (first commit):

remove unused (and inaccurate) LOG_PREFIX from a bunch of places
remove copypased (and inaccurate) comments
fix a couple of comments
    
move `_build_data_section_usage_by_node_with_org_details` & `_build_data_section_usage_by_job` from report base to report_ccsp_v2 (only used there)
    
replace `_init_dimensions` for main sheet + same for loop in each `_build_*` function,
move into base `add_sheet(title, sheet_index, widths=None)`
    
Fail with "No billing data" only when all dataframes are empty
([0] may be None if no sheets are using it even when data is available)

---

Example output:

* https://docs.google.com/spreadsheets/d/1ohWtDmgKOqRFDc1YWk545FRvITSbLg4S/edit?gid=1350376622#gid=1350376622 (last sheet)
* https://docs.google.com/spreadsheets/d/1kDx82n1IVF0T4ZY2Iuz2bDh5fJT9A6mV/edit?gid=1669669102#gid=1669669102 (only sheet)

himdel commented:
```
+ xlsx2csv -a ./metrics_utility/test/test_data//reports/2024/04/CCSPv2-2024-04.xlsx
-------- 1 - Data collection status
Collection timestamp,Filter since,Filter until,CSV filename,Status,Elapsed
2024-04-17 13:59:12.537121+00:00,2024-04-15 00:00:00+00:00,2024-04-16 00:00:00+00:00,job_host_summary.csv,ok,0
2024-04-17 13:59:12.576144+00:00,2024-04-16 00:00:00+00:00,2024-04-17 00:00:00+00:00,job_host_summary.csv,ok,0
2024-04-17 13:59:12.592113+00:00,2024-04-17 00:00:00+00:00,2024-04-17 13:59:12.470584+00:00,job_host_summary.csv,ok,0
```

:tada: :)

AlexSCorey commented on metrics_utility/automation_controller_billing/report/report_ccsp.py:
```diff
@@ -102,70 +94,58 @@ def build_spreadsheet(self):
                 # Sheet with usage by modules
-                self.wb.create_sheet(title='Usage by modules')
-                ws = self.wb.worksheets[sheet_index]
+                ws = self.add_sheet('Usage by modules', sheet_index, self.config['data_column_widths'])
```
We seem to be using this value in many places in this file.  Perhaps it makes sense to assign it to a variable and use that variable in these if blocks.  Might make future maintenance easier.

himdel replied:
It's only used once here, and adding a level of indirection for a string would make things more confusing.
This way, we immediatgely see this is about usage by modules, if we move it to a constant, you'd have to look up the constant every time to tell which sheet it is.

himdel committed:
Report cleanups, initialize_columns -> add_sheet refactor

remove unused (and inaccurate) LOG_PREFIX from a bunch of places
remove copypased (and inaccurate) comments
fix a couple of comments

move _build_data_section_usage_by_node_with_org_details & _build_data_section_usage_by_job from report base to report_ccsp_v2

replace _init_dimensions for main sheet + same for loop in each _build_* function,
move into base add_sheet(title, sheet_index, columns=None)

Fail with "No billing data" only when all dataframes are empty
([0] may be None if no sheets are using it even when data is available)

himdel committed:
Add optional report sheet showing collection coverage

Issue: AAP-41887

AlexSCorey commented on metrics_utility/automation_controller_billing/report/report_ccsp_v2.py:
```diff
@@ -178,8 +182,100 @@ def build_spreadsheet(self):
+
+        return current_row + row_counter
+
+    def _build_data_section_collection_missing(self, current_row, ws, df):
```
It isn't super clear to me what this function is doing in part.  I think the function name might be confusing.  Could we 1) rename the function and/or 2) add more detail to the comments?

AlexSCorey commented on metrics_utility/automation_controller_billing/report/report_ccsp_v2.py:
```diff
@@ -178,8 +182,100 @@ def build_spreadsheet(self):
+        rows = dataframe_to_rows(dataframe, index=False)
+        return self._build_table(current_row, ws, rows)
+
+    def _build_data_section_collection_status(self, first_row, ws, df):
```
To build the rows for this file, are we looping through each row, of each file in the tarball?

himdel replied:
Not each file, just `data_collection_status.csv`

AlexSCorey commented on metrics_utility/automation_controller_billing/report/base.py:
```diff
@@ -17,6 +17,8 @@ class Base:
     RED_COLOR_HEX = 'FF0000'
     LIGHT_BLUE_COLOR_HEX = 'd4eaf3'
     GREEN_COLOR_HEX = '92d050'
+    YELLOW_WARNING_COLOR_HEX = 'FFCC17'
```
I probably did this, but here is a suggestion.
```suggestion
    YELLOW_COLOR_HEX = 'FFCC17'
```
or
```suggestion
WARNING_COLOR_HEX = 'FFCC17'
```

In either case I think we should be consistent in the naming since I expect we will use these colors more as we grow.  Additionally, I think the colors should match the colors and use cases that [Patternfly](https://www.patternfly.org/design-foundations/colors) uses

AlexSCorey reviewed (commented):
All that code clean up feels so good.

AlexSCorey reviewed (commented):
I ran the test and the report was created an looks great.  The code looks good, I've just left a few comments, and a few questions.  

I did have some concerns about the top part of the report.  It is a bit confusing to me and I'm not sure about its purpose.
![Screenshot 2025-04-15 at 4 15 24 PM](https://github.com/user-attachments/assets/5c36812d-ce2b-4fa2-afdb-78a2b7523fde)

himdel committed:
review

himdel commented:
Thanks @AlexSCorey, added a comment on the "*missing" function, and fixed the color variable :)

AlexSCorey reviewed (approved):
This looks good to me.  Still a bit concerned about the code coverage reporting.
